src/bin/pg_autoctl/supervisor.c

```diff
@@ -56,22 +74,278 @@ static bool supervisor_restart_service(Supervisor *supervisor,
+ * recentlyRemoved array and we will hold it there until it is requested via
+ * supervisor_dynamic_remove_terminated_service() which subsequently removes it.
+ *
+ * Returns true if the service is stopped and removed from the dynamicService
+ * array, false otherwise.
+ */
+bool
+supervisor_dynamic_service_disable(Supervisor *supervisor,
+								   const char *serviceName)
+{
+	Service *dynamicService = NULL;
+	int serviceCount = supervisor->dynamicServicesEnabled.serviceCount;
+	int serviceIndex;
+
+	/* Find it in the dynamic array */
+	for (serviceIndex = 0; serviceIndex < serviceCount; serviceIndex++)
+	{
+		Service *service = &supervisor->dynamicServicesEnabled.services[serviceIndex];
+
+		if (!strncmp(service->name, serviceName, strlen(serviceName)))
+		{
+			dynamicService = service;
+			break;
+		}
+	}
+
+	/* It was not found, nothing to do. */
+	if (!dynamicService)
+	{
+		return false;
+	}
+
+	/*
+	 * Since we are going to signal the service, we have to make certain that we
+	 * will not error out in the supervisor_loop on the waitpid() response. So
+	 * after killing the service and removing it from the dynamicService array,
+	 * we add it to the recentlyRemoved array to find it again later.
+	 */
+	if (!kill(dynamicService->pid, SIGTERM))
+	{
+		log_info("Service %s with %d did not receive signal",
+				 dynamicService->name, dynamicService->pid);
+	}
+
+	log_info("Stopped pg_autoctl %s service with pid %d",
+			 dynamicService->name, dynamicService->pid);
+
+	/*
+	 * Remove the dynamic Service from the array by swapping it out with the
+	 * last element in the array and bringing the array count down by one
+	 */
+	supervisor->dynamicServicesEnabled.services[serviceIndex] =
+		supervisor->dynamicServicesEnabled.services[serviceCount - 1];
+	supervisor->dynamicServicesEnabled.serviceCount--;
+
+	/*
+	 * Finally add the dynamic service to the recently removed array.
+	 *
+	 * If we reached the end of this array, start again at the beginning. This
+	 * is a loosely held array and we do not fuss too much.
+	 */
+	if (supervisor->dynamicServicesDisabled.serviceCount == MAXDYNSERVICES)
+	{
+		supervisor->dynamicServicesDisabled.serviceCount = 0;
+	}
+
+	/*
+	 * IGNORE-BANNED fits here because we want to take ownership of the service.
+	 * We copy to an address that we own from a non overlapping address.
+	 */
+	memcpy(&supervisor->dynamicServicesDisabled.services[    /* IGNORE-BANNED */
+			   supervisor->dynamicServicesDisabled.serviceCount],
+		   dynamicService,
+		   sizeof(*dynamicService));
+	supervisor->dynamicServicesDisabled.serviceCount++;
+
+	return true;
+}
+
+
 /*
  * supervisor_start starts given services as sub-processes and then supervise
  * them.
  */
 bool
-supervisor_start(Service services[], int serviceCount, const char *pidfile)
+supervisor_start(Service services[], int serviceCount, const char *pidfile,
+				 void (*dynamicHandler)(Supervisor *, void *, int *),
+				 void *dynamicHandlerArg)
```

Comment:
I would disagree with you. I am sorry about it.

I tried the hardcoded way initially and is by far much more invasive and forces one to jump through hoops in many parts of the code (all the way down to config) in order to avoid simply adding a function pointer. I do not see it.

Having said that, this is your project and you are the maintainer. I am simply the code monkey here. My goal is to get this in, so I will do what you say, but will kindly ask for someone to actually try it out and have the discussion there. If possible of course.

Reply:
I'm surprised. In my mind I was seeing the same code block just in a different place really. The thing you tried, was it with an extra dynamic service array as input to the `supervisor_start` function, and the idea that the function (or the main loop later) would loop through the array and call `supervisor_enable_dynamic_service` with the service definition as input?

Reply:
Correct. You might want to think about the cases where you want to enable a service and add it an already running process group several days after the process group has started running. Same goes for disabling services.

You can do all that with at least one array more and some more code in other parts of the code. Or with keeping that information in the Service itself. An idea initially presented to you and shot down if you remember.

What am I missing?

Reply:
So I think this discussion is hard to have on text only medium, and I am open for a live chat if you're available. I wanted to help drive this PR into being a quick first commit from you, but apparently my optimism gets in the way of a proper review with enough time spent thinking... sorry about that. 

The design I had in mind, and you're right, that does not work, is that the main supervisor loop uses the two arrays of dynamic services (enabled, disabled) and always process them in the same way. When a signal is received (SIGHUP or SIGUSR1) that the setup might have changed, the main loop would review those arrays and start or stop dynamic services.

This doesn't work because at the moment in the PR the service definition is lazily shared with the supervisor entity. That said, it is a static definition: the only part that is dynamic is the enabled/disabled information, and that's user-driven.

We also have the problem that the SIGUSR1 signal is currently hard-coded to have the supervisor process itself read the keeper configuration file again and update its own keeper instance. I think that's why you have a modularity violation, that you are solving with a function pointer and opaque `void *` pointer for a state. This also brings the question, what if we enable pgbouncer on the monitor? Then the supervisor also needs to maintain a Monitor struct instance in-memory. Etc.

The central question here is: how the keeper node-active process communicates with the supervisor that a known dynamic service has been enabled/disabled? 

That's the whole reason why we have this [service_postgres_ctl.c](https://github.com/citusdata/pg_auto_failover/blob/master/src/bin/pg_autoctl/service_postgres_ctl.c) intermediate process, and the communication is done through a specific state file implemented at https://github.com/citusdata/pg_auto_failover/blob/master/src/bin/pg_autoctl/state.h#L158 with the `ExpectedPostgresStatus` struct. The client process (node-active or listener) can then edit the on-disk state file with a new enum value, and the Postgres controller process that loops over reading that file every 100ms then implements the change.

I am now thinking (out loud) that we need the same thing for each dynamic service we want to run: a controller process (always running) that monitors the process expected state file (enabled/disabled in the pgbouncer case, and we could add the pause/suspend/resume bits in there) and implements the change.

With that, the keeper config file is owned by a single process and code path, the keeper node-active service. If we want to implement pgbouncer in the monitor, we can handle the monitor setup in the monitor code-path and process and have the monitor signal to the pgbouncer controller process that the expected state of the pgbouncer process has changed, but just writing new content to a (possibly new) state file.

It means that the supervisor doesn't have to handle the enabled/disabled state if we go with this design. It does not need to handle dynamic services at all, actually. It's all static controller processes that are always running and might be asked to start a service at some point.

I'm sorry that I am only reaching that conclusion now, and that we needed so much work to get there. What are your thoughts?

Reply:
Very eloquent description. Kudos for that.

I would disagree on a couple of points but I agree on most. Let us take the discussion elsewhere and return with our conclusions here.

And yes, it takes a while to reach a conclusion. No need for apologies :)